### ISSUE TITLE
add cdist function to AbstractArray

### ISSUE THREAD

**Issue by a-quelle-pasqal** [pasqal-io/Pulser#927]
We currently have the following piece of code in `HamiltonianData`

```
if not positions[0].is_tensor:
            return cast(
                np.ndarray,
                np.round(
                    cast(
                        np.ndarray,
                        cdist(positions, positions, metric="euclidean"),
                    ),
                    COORD_PRECISION,
                ),
            )
        else:
            ten = pm.torch.stack([x._array for x in positions])
            return pm.torch.cdist(ten, ten)
```

here `positions` is a `list[AbstractArray]`. Let's have a function `cdist(list[AbstractArray], list[AbstractArray])` that does the following.

**Comment by danielgaskins** [comment]
Hi, I'm happy to handle this one. I confirmed that the backend-specific distance logic is still present on `develop`. I plan to add `pulser.math.cdist` with NumPy and Torch support, cover both paths including gradient preservation, and update `HamiltonianData` to use it. Could you assign this issue to me?

**Comment by a-corni** [comment]
Hi @danielgaskins !
Great! I have assigned you the issue.
